ui/qt-lib/src/Ariadne/UI/Qt/Widgets/Dialogs/Send.hs

```diff
@@ -0,0 +1,252 @@
+module Ariadne.UI.Qt.Widgets.Dialogs.Send
+  ( SendInputs(..)
+  , SendOptions(..)
+  , SendResult(..)
+  , runSend
+  ) where
+
+import qualified Data.Text as T
+import Data.Scientific
+
+import Graphics.UI.Qtah.Signal (connect_)
+
+import Graphics.UI.Qtah.Core.HPoint (HPoint(..))
+import Graphics.UI.Qtah.Core.HSize (HSize(..))
+import qualified Graphics.UI.Qtah.Core.QEvent as QEvent
+import qualified Graphics.UI.Qtah.Core.QObject as QObject
+import qualified Graphics.UI.Qtah.Event as Event
+import qualified Graphics.UI.Qtah.Gui.QMouseEvent as QMouseEvent
+import qualified Graphics.UI.Qtah.Widgets.QAbstractButton as QAbstractButton
+import qualified Graphics.UI.Qtah.Widgets.QBoxLayout as QBoxLayout
+import qualified Graphics.UI.Qtah.Widgets.QCheckBox as QCheckBox
+import qualified Graphics.UI.Qtah.Widgets.QDialog as QDialog
+import qualified Graphics.UI.Qtah.Widgets.QHBoxLayout as QHBoxLayout
+import qualified Graphics.UI.Qtah.Widgets.QLabel as QLabel
+import qualified Graphics.UI.Qtah.Widgets.QLineEdit as QLineEdit
+import qualified Graphics.UI.Qtah.Widgets.QPushButton as QPushButton
+import qualified Graphics.UI.Qtah.Widgets.QVBoxLayout as QVBoxLayout
+import qualified Graphics.UI.Qtah.Widgets.QWidget as QWidget
+
+import Ariadne.UI.Qt.Face
+import Ariadne.UI.Qt.UI
+import Ariadne.UI.Qt.Util
+import Ariadne.UI.Qt.Widgets.Dialogs.Util
+
+data SendInputs = SendInputsSingle Word | SendInputsMulti [UiAccountInfo]
+
+data SendOptions =
+  SendOptions
+    { soAddress :: Text
+    , soAmount :: Scientific
+    , soAccounts :: [Word]
+    }
+
+data SendResult = SendSuccess SendOptions | SendCancel
+
+data AccountCheckbox =
+  AccountCheckbox
+    { aIdx :: Word
+    , checkbox :: QCheckBox.QCheckBox
+    , accountName :: Text
+    }
+
+data AccountSelector =
+  AccountSelector
+    { dropdownWidget :: QWidget.QWidget
+    , checkboxes :: [AccountCheckbox]
+    }
+
+data Send =
+  Send
+    { send :: QDialog.QDialog
+    , amountEdit :: QLineEdit.QLineEdit
+    , addressEdit :: QLineEdit.QLineEdit
+    , sendButton :: QPushButton.QPushButton
+    , fromDisplay :: QLabel.QLabel
+    , accountSelector :: Either Word AccountSelector
+    }
+
+initSend :: SendInputs -> IO Send
+initSend sendInputs = do
+  send <- QDialog.new
+  layout <- createLayout send
+
+  QWidget.setWindowTitle send ("SEND" :: String)
+
+  amountLabel <- QLabel.newWithText ("<b>AMOUNT</b>" :: String)
+  amountLayout <- QHBoxLayout.new
+  amountEdit <- QLineEdit.new
```

Comment:
I can't use `QDoubleSpinBox` since it leads to one of two possible problems:

- If I get the value with `QDoubleSpinBox.value`, it's a `Double` and conversion to Scientific is unsafe theoretically
- If I get the value with `QDoubleSpinBox.text`, it's a string that depends on user's locale -- for me it has `,` as decimal separator, and we can't feed that string into `readMaybe` / `readEither`.

This however means that we must validate the value ourselves. And AFAIU we can't use `QDoubleValidator` for the same reason above and we can't subclass `QValidator` to make our own one.

Reply:
> And AFAIU we can't use QDoubleValidator for the same reason

Wait, why not? AFAIU, `QDoubleValidator` is not really about `double` type.

> QDoubleValidator provides an upper bound, a lower bound, and a limit on the number of digits after the decimal point.

It doesn't force us to get `Double`, does it?

Reply:
It doesn't. But it's still affected by locale, so will accept only commas. I can double check, but I'm pretty sure that's how it works.

Reply:
Kostylee
Anyway, I fear that is inconvenient for the user too. I, for example, find this locale setting stupid and would be irritated by the need to enter the comma. And would not understand what's going on when I can enter dot.

Reply:
What we __can__ do however, is to create the validator with `C` locale it whatever. What do you think?

Reply:
Is there some method to set locale for a validator? Sounds good to me then.

Reply:
I had to add locales to qtah. Now it seems to work as intended.
Also please note that I had to disable group separators for locale. `C` locale has `,` as group separator, but `read` won't parse it.

Reply:
What is group separator?